_posts/2021-01-15-tuples-bring-generic-programming-to-scala-3.md

```diff
@@ -0,0 +1,326 @@
+```
+
+If you are implementing an operation such as serializing instances of these
+types to CSV or JSON, you will realize that the logic is exactly the same and
+you will want to implement it only once. This is equivalent on defining the
+serialization algorithm for the `(String, Int, Boolean)` HList, assuming that
+you can reduce both case classes to it.
+
+# A simple CSV encoder
+
+Let's consider a simple CSV encoder for our Employee and IceCream case classes.
+Each record, or line, of a CSV file is a sequence of values separated by a
+delimiter, usually a comma or a semicolon. In Scala we can represent each value
+as text, using the `String` type, and thus each record can be a list of values,
+with type `List[String]`. Therefore, in order to encode case classes to CSV, we
+need to extract each field of the case class and to turn it into a `String`,
+before collecting all the fields in a list.  In this setting, `Employee` and
+`IceCream` could be treated in the same way, because they can be simply be seen
+as a `(String, Int, Boolean)` which need to be transformed into a
+`List[String]`.  We will first see how to handle this simple scenario before
+briefly looking at how to obtain a tuple from a case class.
+
+Assuming that we know how to transform each element of a tuple into a
+`List[String]`, can we transform any tuple into a `List[String]` ?
+
+The answer is yes, and this is possible because Scala 3 introduces types `*:`,
+`EmptyTuple` and `NonEmptyTuple` but also methods `head` and `tail` which allow
+us to define recursive operations on tuples.
+
+## Set up
+
+Let's define the `Enc[A]` type-class, which describes the capability of values
+of type `A` to be converted into `List[String]`:
+
+```scala
+trait Enc[A]:
+  def toCsv(a: A): List[String]
+```
+
+We can then add some instances for our base types:
+
+```scala
+object BaseEnc:
+  given Enc[Int] with
+    def toCsv(x: Int) = List(x.toString)
+
+  given Enc[Boolean] with
+    def toCsv(x: Boolean) = List(if x then "true" else "false")
+
+  given Enc[String] with
+    def toCsv(x: String) = List(x)
+```
+
+## Recursion!
+
+Now that all these tools are in place, let's focus on the hard part:
+implementing the actual transformation. Similarly to how you may be used to
+recurse on lists, on tuples we need to manage two scenarios: the base case
+(`EmptyTuple`) and the inductive case (`NonEmptyTuple`).
+
+In the following snippet, I prefer to use the [context bound
+syntax](https://dotty.epfl.ch/docs/reference/contextual/context-bounds.html)
+even if I need a handle for the instances because it concentrates all the
+constraints in the type parameter list (and I do not need to come up with any
+name).  After this personal preference disclaimer, let's see the two cases:
+
+```scala
+object TupleEnc:
+  // Base case
+  given [T: Enc]: Enc[T *: EmptyTuple] with
+    def toCsv(oneElement: T *: EmptyTuple) =
+      summon[Enc[T]].toCsv(oneElement.head)
```

Comment:
For reference, with the `ColumnEncoder`/`RowEncoder` suggestion, this given definition would have the following signature:
```suggestion
  given [T: ColumnEncoder]: RowEncoder[T *: EmptyTuple] with
    def encode(oneElement: T *: EmptyTuple) =
      List(summon[ColumnEncoder[T]].encode(oneElement.head))
```